pds_doi_core/util/test/doi_validator_test.py

```diff
@@ -0,0 +1,427 @@
+
+logger = get_logger(__name__)
+
+# This function remove_test_artifact() is outside of the class to allow access to the removal of a test artifact.
+def remove_test_artifact(db_name):
+    if os.path.isfile(db_name):
+        os.remove(db_name)
+        logger.info(f"Removed test artifact database file {db_name}")
+    else:
+        logger.info(f"File not exist, test artifact database file {db_name}")
+
+class MyTestCase(unittest.TestCase):
+
+    # A note about the names of the tests:
+    #
+    #     For some strange reason, unittest module run the tests in alphabetical order.
+    #     Since we wish to run the test in the order in this file, they have added token 'AAA' so the alphabetical order is the same as the sequence of their appearance.
+
+    db_name = 'doi_temp_for_doi_validator_unit_test.db'  # This file is removed in the beginning and at the end of these tests.
+
+    @classmethod
+    def set_initial_states(self):
+        logger.info("RUNNING_TEST: set_initial_states")
+
+        # This function should be called from the first test to remove any test artifacts and set up some records.
+
+        self._database_obj = DOIDataBase(self.db_name)
+        self._doi_validator = DOIValidator(db_name=self.db_name)
+
+        # This should be done before any testing.
+        remove_test_artifact(self.db_name)
+
+        # Use the setUp() function to initializes many fields and write the initial record.
+        self.setUp()
+
+        # Write a record into database
+        # The value of 'doi', 'lid', 'vid' fields are None.  Other later calls write_doi_info_to_database() to write other records with valid 'doi' field.
+        # (lid,vid,transaction_key,doi,transaction_date,status,title,product_type,product_type_specific,submitter,discipline_node)
+        self._database_obj.write_doi_info_to_database(None, None, self.transaction_key, None, self.transaction_date,
+                                                      self.status, self.title, self.product_type, self.product_type_specific,
+                                                      self.submitter, self.discipline_node)
+
+        logger.info("Writing a record to database file {self.db_name}.")
+
+    @classmethod
+    def setUp(self):
+        # This setUp() function is called for every test.
+        logger.info(f"setUp() called")
+
+        self.lid = 'urn:nasa:pds:lab_shocked_feldspars'
+        self.vid = '1.0'
+        self.transaction_key = './transaction_history/img/2020-06-15T18:42:45.653317'
+        self.transaction_date = datetime.datetime.now()
+        self.status = 'Draft'
+        self.title = 'Laboratory Shocked Feldspars Bundle'
+        self.product_type = 'Collection'
+        self.product_type_specific = 'PDS4 Collection'
+        self.submitter = 'Qui.T.Chau@jpl.nasa.gov'
+        self.discipline_node = 'img'
+
+        self.id = '21940'
+        self.doi = '10.17189/' + self.id
+
+    def test_AAA_draft_001(self):
```

Comment:
That is not good to have numbering AAA or 001 in method. This should be more descriptive even for testing methods.

Reply:
From what I have observed, inorder for the tests to run in the order specified, their names have to have some ordering otherwise weird errors are occurring because some conditions are not met.   The tests have to run in the order specified in the file.  

Not quite sure how to response to your desire not to see numbers or special token in the names.

I tested by changing the first draft test test_AAA_draft_001 to something else and the whole test failed:

    #def test_AAA_draft_001(self):
    def test_draft_with_existing_title_no_doi_minted(self):

{pds-dev-el7.jpl.nasa.gov}/home/qchau/sandbox/pds-doi-service 237 % python3 pds_doi_core/util/test/doi_validator_test.py

======================================================================
ERROR: test_AAA_draft_002 (__main__.MyTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pds_doi_core/util/test/doi_validator_test.py", line 124, in test_AAA_draft_002
    self._doi_validator.validate(doi_obj,action_name)
AttributeError: 'MyTestCase' object has no attribute '_doi_validator'

======================================================================
ERROR: test_AAA_draft_003 (__main__.MyTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pds_doi_core/util/test/doi_validator_test.py", line 141, in test_AAA_draft_003
    self._database_obj.write_doi_info_to_database(self.lid, self.vid, self.transaction_key, self.doi, self.transaction_date,
AttributeError: 'MyTestCase' object has no attribute '_database_obj'

======================================================================
ERROR: test_AAA_draft_004 (__main__.MyTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pds_doi_core/util/test/doi_validator_test.py", line 165, in test_AAA_draft_004
    self.assertRaises(DuplicatedTitleDOIException,self._doi_validator.validate,doi_obj,action_name)
AttributeError: 'MyTestCase' object has no attribute '_doi_validator'

======================================================================
ERROR: test_AAA_reserve_001 (__main__.MyTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "pds_doi_core/util/test/doi_validator_test.py", line 182, in test_AAA_reserve_001
    self.assertRaises(DuplicatedTitleDOIException,self._doi_validator.validate,doi_obj,action_name)
AttributeError: 'MyTestCase' object has no attribute '_doi_validator'

----------------------------------------------------------------------
Ran 17 tests in 0.421s

FAILED (errors=4)

Reply:
Hi Qui,
You are right if the order is important that make sense.
However I don't think we should count on the naming of the method to have an implicit order.
I quickly looked on google and I like the suggestion here (https://stackoverflow.com/questions/20711300/controlling-execution-order-of-unit-tests-in-visual-studio).

They advise to embedded the single tests into a method which will call the unit tests which have to be run into a specific order.
An alternative would be to make the unit test independent from the order and redefine an appropriate test database for each but this is more work and we don't need that.

Reply:
Combining all these tests as private function name with underscore '_' and then having a big function that starts with test that calls all the private functions worked.

It only shows 2 tests ran though but all the tests got ran:

{pds-dev-el7.jpl.nasa.gov}/home/qchau/sandbox/pds-doi-service 246 % python3 pds_doi_core/util/test/doi_validator_test.py

Ran 2 tests in 0.305s

OK


{pds-dev-el7.jpl.nasa.gov}/home/qchau/sandbox/pds-doi-service 245 % python3 pds_doi_core/util/test/doi_validator_test.py >& t1; grep RUNNING_TEST t1
INFO __main__:_test_draft_existing_title_no_doi RUNNING_TEST:_test_draft_existing_title_no_doi
INFO __main__:set_initial_states RUNNING_TEST: set_initial_states
INFO __main__:_test_draft_existing_title_existing_doi RUNNING_TEST:_test_draft_existing_title_existing_doi
INFO __main__:_test_draft_existing_title_existing_doi_duplicate_exception RUNNING_TEST:_test_draft_existing_title_existing_doi_duplicate_exception
INFO __main__:_test_reserve_existing_title_existing_doi_duplicate_exception RUNNING_TEST:_test_reserve_existing_title_existing_doi_duplicate_exception
INFO __main__:_test_reserve_existing_title_no_doi RUNNING_TEST:_test_reserve_existing_title_no_doi
INFO __main__:set_initial_states RUNNING_TEST: set_initial_states
INFO __main__:_test_reserve_existing_title_existing_doi_duplicate_title RUNNING_TEST:_test_reserve_existing_title_existing_doi_duplicate_title
INFO __main__:_test_reserve_existing_title_existing_doi_existing_lidvid RUNNING_TEST:_test_reserve_existing_title_existing_doi_existing_lidvid
INFO __main__:_test_reserve_new_title_existing_doi_new_lidvid RUNNING_TEST:_test_reserve_new_title_existing_doi_new_lidvid
INFO __main__:_test_reserve_new_title_existing_doi_existing_lidvid RUNNING_TEST:_test_reserve_new_title_existing_doi_existing_lidvid
INFO __main__:_test_reserve_existing_title_different_doi_existing_lidvid RUNNING_TEST:_test_reserve_existing_title_different_doi_existing_lidvid
INFO __main__:_test_release_existing_title_existing_doi_existing_lidvid RUNNING_TEST:_test_release_existing_title_existing_doi_existing_lidvid
INFO __main__:_test_release_new_title_existing_doi_existing_lidvid RUNNING_TEST:_test_release_new_title_existing_doi_existing_lidvid
INFO __main__:_test_release_new_title_existing_doi_new_lidvid RUNNING_TEST:_test_release_new_title_existing_doi_new_lidvid
INFO __main__:_test_last RUNNING_TEST:_test_last
INFO __main__:test_release_existing_title_new_doi_existing_lidvid RUNNING_TEST:_test_release_new_title_existing_doi_new_lidvid
{pds-dev-el7.jpl.nasa.gov}/home/qchau/sandbox/pds-doi-service 246 %